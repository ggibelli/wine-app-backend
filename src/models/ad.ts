/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import mongoose, { Schema, Document } from 'mongoose';
// import { Address } from '../types';
import {
  // Province,
  // Regioni,
  METODOPRODUTTIVO,
} from '../utils/enumMongooseHelper';
import { IUserDoc } from './user';
import { IWineDoc } from './wine';
import { TypeAd, TypeProduct, Menzione } from '../types';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { Negotiation } from '../generated/graphql';

export interface IAd {
  postedBy: IUserDoc['_id'];
  wineName?: string;
  wine?: IWineDoc['_id'];
  sottoZona?: string;
  menzione?: Menzione;
  metodoProduttivo?: MetodoProduttivo;
  vineyardName?: string;
  vineyard?: IVineyardDoc['_id'];
  harvest: number;
  abv: number;
  priceFrom: number; //se vendita priceFrom e priceTo settati a stesso numero
  priceTo: number;
  litersFrom?: number; //se vendite litersFrom e litersTo settati a stesso numero
  litersTo?: number;
  kgFrom?: number;
  kgTo?: number;
  content?: string | null | undefined;
  // address: Address;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'annuncio, solo attive, solo graphql??
  viewedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  savedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  typeAd: TypeAd;
  typeProduct: TypeProduct;
  needsFollowUp: boolean;
  isActive: boolean;
  datePosted: Date;
}

export interface AdGraphQl extends IAd {
  _id: mongoose.Types.ObjectId | string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const adSchemaFields: Record<keyof IAd, any> = {
  postedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  typeAd: {
    type: String,
    enum: [TypeAd.SELL, TypeAd.BUY],
    required: true,
  },
  typeProduct: {
    type: String,
    enum: [TypeProduct.ADWINE, TypeProduct.ADGRAPE],
    required: true,
  },
  wineName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
    index: true,
  },
  wine: {
    type: Schema.Types.ObjectId,
    ref: 'Wine',
    // required() {
    //   return this.typeProduct === TypeProduct.ADWINE;
    // },
  },
  sottoZona: {
    type: String,
  },
  menzione: {
    type: String,
    enum: ['CLASSICO', 'RISERVA', 'SUPERIORE', 'VIGNA'],
  },
  metodoProduttivo: {
    type: String,
    enum: Object.values(METODOPRODUTTIVO),
    required: true,
    default: 'CONVENZIONALE',
  },
  vineyardName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
    index: true,
  },
  vineyard: {
    type: Schema.Types.ObjectId,
    ref: 'Vineyard',
    // required() {
    //   return this.typeProduct === TypeProduct.ADGRAPE;
    // },
  },
  abv: Number,
  harvest: {
    type: Number,
    required: true,
  },
  priceFrom: {
    type: Number,
    required: true,
  },
  priceTo: {
    type: Number,
    required: true,
  },
  litersFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  litersTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  kgFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  kgTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  // address: {
  //   comune: {
  //     type: String,
  //     required: true,
  //   },
  //   provincia: {
  //     type: String,
  //     enum: Object.values(Province),
  //     required: true,
  //   },
  //   regione: {
  //     type: String,
  //     enum: Object.values(Regioni),
  //     required: true,
  //   },
  // },
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  viewedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  savedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  content: {
    type: String,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  needsFollowUp: {
    type: Boolean,
    default: true,
  },
  datePosted: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

export interface IAdDoc extends IAd, Document {}

const adSchema = new Schema(adSchemaFields);

adSchema.index({ typeProduct: 1, typeAd: 1, wineName: 1, vineyardName: 1 });

export const Ad = mongoose.model<IAdDoc>('Ad', adSchema);
