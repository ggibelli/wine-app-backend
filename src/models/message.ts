import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { INegotiationDoc } from './negotiation';

export interface IMessage {
  negotiation: INegotiationDoc['_id'];
  content: string;
  sentBy: IUserDoc['_id'];
  sentTo: IUserDoc['_id'];
  dateSent: Date;
  isViewed: boolean;
}

export interface MessageGraphQl extends IMessage {
  _id: mongoose.Types.ObjectId;
}

export interface IMessageDoc extends IMessage, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const messageSchemaFields: Record<keyof IMessage, any> = {
  negotiation: {
    type: Schema.Types.ObjectId,
    ref: 'Negotiation',
    required: true,
  },
  content: {
    type: String,
    minlength: 1,
    required: true,
  },
  sentBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  sentTo: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  dateSent: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  isViewed: {
    type: Boolean,
    default: false,
  },
};

const messageSchema = new Schema(messageSchemaFields);

messageSchema.index({ negotiation: 1, sentTo: 1, sentFrom: 1 });

export const Message = mongoose.model<IMessageDoc>('Message', messageSchema);
