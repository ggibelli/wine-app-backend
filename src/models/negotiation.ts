import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { IAdDoc } from './ad';
import { IMessageDoc } from './message';
import { TypeAd } from '../types';
import mongooseUniqueValidator from 'mongoose-unique-validator';

export interface INegotiation {
  createdBy: IUserDoc['_id'];
  ad: IAdDoc['_id'];
  forUserAd: IUserDoc['_id'];
  messages?: mongoose.Types.Array<IMessageDoc['_id']>;
  isConcluded: boolean;
  dateCreated: Date;
  type: TypeAd;
}

export interface NegotiationGraphQl extends INegotiation {
  _id: mongoose.Types.ObjectId;
}

export interface INegotiationDoc extends INegotiation, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const negotiationSchemaFields: Record<keyof INegotiation, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  type: {
    type: String,
    enum: ['SELL', 'BUY'],
    required: true,
  },
  ad: {
    type: Schema.Types.ObjectId,
    ref: 'Ad',
    required: true,
  },
  forUserAd: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  isConcluded: {
    type: Boolean,
    default: false,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

const negotiationSchema = new Schema(negotiationSchemaFields);

negotiationSchema.index({ createdBy: 1, ad: 1 }, { unique: true });
negotiationSchema.plugin(mongooseUniqueValidator);
export const Negotiation = mongoose.model<INegotiationDoc>(
  'Negotiation',
  negotiationSchema
);
