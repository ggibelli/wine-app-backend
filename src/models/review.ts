import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { INegotiationDoc } from './negotiation';
import { TypeAd, Rating } from '../types';
import mongooseUniqueValidator from 'mongoose-unique-validator';

export interface IReview {
  createdBy: IUserDoc['_id'];
  negotiation: INegotiationDoc['_id'];
  forUser: IUserDoc['_id'];
  rating: Rating;
  dateCreated: Date;
  content: string;
  type: TypeAd;
}

export interface ReviewGraphQl extends IReview {
  _id: mongoose.Types.ObjectId;
}

export interface IReviewDoc extends IReview, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const reviewSchemaFields: Record<keyof IReview, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  negotiation: {
    type: Schema.Types.ObjectId,
    ref: 'Negotiation',
    required: true,
  },
  forUser: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  content: {
    type: String,
    minlength: 5,
  },
  rating: {
    type: String,
    enum: ['POOR', 'AVERAGE', 'OK', 'GOOD', 'PERFECT'],
    required: true,
  },
  type: {
    type: String,
    enum: ['SELL', 'BUY'],
  },
};

const reviewSchema = new Schema(reviewSchemaFields);

reviewSchema.index({ createdBy: 1, negotiation: 1 }, { unique: true });
reviewSchema.plugin(mongooseUniqueValidator);

export const Review = mongoose.model<IReviewDoc>('Review', reviewSchema);
