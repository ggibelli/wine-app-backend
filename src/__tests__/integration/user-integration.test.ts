/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { gql } from 'apollo-server-express';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../../tests/integrationSetup';
import { User } from '../../models/user';
import { ads, users } from '../../tests/mocksTests';
import { UserInput } from '../../generated/graphql';
import { Province, Rating, Regioni } from '../../types';
import supertest from 'supertest';
import { app } from '../../index';
import { Ad } from '../../models/ad';
import { Message } from '../../models/message';
import { Negotiation } from '../../models/negotiation';
import { Review } from '../../models/review';
import { Vineyard } from '../../models/vineyard';
import { Wine } from '../../models/wine';
import createVineyardDb from '../../utils/vineyardExtractor';
// import createWineDb from '../../utils/wineExtractor';
import cron from 'cron';

const fakeStart = jest.fn(() => fakeStop());
const fakeStop = jest.fn(() => null);
const FakeCron = jest.fn(() => ({
  stop: fakeStop,
  start: fakeStart,
  fireOnTick: fakeStart,
}));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
cron.CronJob = FakeCron;

const api = supertest(app);
const { query, mutate, setOptions } = testClient;

interface Response {
  data: any;
  errors: any;
}

const LOGIN_VALID = gql`
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
      }
    }
  }
`;

const LOGIN_VALID_OTHER = gql`
  mutation {
    login(email: "gio2@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_INVALID = gql`
  mutation {
    login(email: "wrong", password: "giovanni") {
      response {
        token
      }
      errors {
        name
      }
    }
  }
`;

const CREATE_USER = gql`
  mutation CreateUser($auth: UserInput!) {
    createUser(user: $auth) {
      response {
        token
        user {
          isVerified
        }
      }
      errors {
        name
        text
      }
    }
  }
`;

const ME = gql`
  {
    me {
      pIva
      phoneNumber
      isVerified
    }
  }
`;

const ME_FULL = gql`
  {
    me {
      pIva
      phoneNumber
      isVerified
      ads {
        postedBy {
          firstName
        }
        ... on AdWine {
          wineName
        }
      }
      negotiations {
        createdBy {
          firstName
        }
        forUserAd {
          firstName
        }
      }
      messages {
        content
        sentBy {
          firstName
        }
        sentTo {
          firstName
        }
      }
      reviews {
        createdBy {
          firstName
        }
        forUser {
          firstName
        }
        rating
      }
    }
  }
`;

const USERS = gql`
  {
    users {
      firstName
      phoneNumber
      email
      ads {
        postedBy {
          firstName
        }
        ... on AdWine {
          wineName
        }
      }
      negotiations {
        createdBy {
          firstName
        }
        forUserAd {
          firstName
        }
      }
      messages {
        content
        sentBy {
          firstName
        }
        sentTo {
          firstName
        }
      }
      reviews {
        createdBy {
          firstName
        }
        forUser {
          firstName
        }
        rating
      }
    }
  }
`;

const USER = gql`
  query GetUser($id: ID!) {
    user(id: $id) {
      firstName
      ads {
        postedBy {
          firstName
        }
        ... on AdWine {
          wineName
        }
      }
      negotiations {
        createdBy {
          firstName
        }
        forUserAd {
          firstName
        }
      }
      messages {
        content
        sentBy {
          firstName
        }
        sentTo {
          firstName
        }
      }
      reviews {
        createdBy {
          firstName
        }
        forUser {
          firstName
        }
        rating
      }
    }
  }
`;

const indirizzo = {
  via: 'asd asddasd',
  CAP: '12345',
  comune: 'aaaaa',
  provincia: Province.AT,
  regione: Regioni.PIEMONTE,
};

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const adsMock = ads();
  // const wineDb = await createWineDb();
  const vineyardDb = createVineyardDb();
  const user = new User(usersMock[0]);
  const otherUser = new User(usersMock[1]);
  const thirdUser = new User(usersMock[2]);

  const ad = new Ad({ ...adsMock[0], postedBy: user });
  const otherAd = new Ad({ ...adsMock[1], postedBy: otherUser });
  const negotiation = new Negotiation({
    createdBy: user,
    ad: otherAd,
    forUserAd: otherAd.postedBy,
    type: otherAd.typeAd,
  });
  const otherNegotiation = new Negotiation({
    createdBy: thirdUser,
    ad: ad,
    forUserAd: ad.postedBy,
    type: ad.typeAd,
  });
  const review = new Review({
    createdBy: user,
    negotiation: negotiation,
    forUser: negotiation.forUserAd,
    type: ad.typeAd,
    rating: Rating.PERFECT,
    content: 'perfect',
  });
  const otherReview = new Review({
    createdBy: thirdUser,
    negotiation: otherNegotiation,
    forUser: otherNegotiation.forUserAd,
    type: ad.typeAd,
    rating: Rating.POOR,
    content: 'very poor',
  });
  const vineyard = new Vineyard(vineyardDb[0]);
  const wine = new Wine({
    denominazioneVino: 'Abruzzo',
    espressioneComunitaria: 'DOP',
    denominazioneZona: 'DOC',
    regione: [Regioni.ABRUZZO],
  });
  const message = new Message({
    negotiation: negotiation,
    sentBy: negotiation.createdBy,
    sentTo: negotiation.forUserAd,
    content: 'ciao',
  });
  const otherMessage = new Message({
    negotiation: otherNegotiation,
    sentBy: otherNegotiation.createdBy,
    sentTo: otherNegotiation.forUserAd,
    content: 'ciao di nuovo',
  });
  await message.save();
  await otherMessage.save();
  await otherUser.save();
  await thirdUser.save();
  await ad.save();
  await otherAd.save();
  await negotiation.save();
  await otherNegotiation.save();
  await review.save();
  await otherReview.save();
  await vineyard.save();
  await wine.save();

  await user.save();
});

describe('Integration test users', () => {
  it('valid login mutation returns token', async () => {
    const res: Response = await query(LOGIN_VALID);
    //expect(res).toMatchSnapshot();
    expect(res.data.login.response.token).toBeDefined();
  });

  it('invalid login mutation returns errors', async () => {
    const res = await query(LOGIN_INVALID);
    expect(res).toMatchSnapshot();
    //expect(res.data.login.response).toBeNull();
    //expect(res.data.login.errors).toBeDefined();
  });

  it('create user mutation successfull and activation successfull', async () => {
    jest.setTimeout(30000);

    const auth: UserInput = {
      email: 'giollo@ollo.it',
      password: 'Pro@3###4~~vaPassWord!23',
      firstName: 'Giova',
      lastName: 'Gibelli',
      pIva: '01640250054',
      phoneNumber: '1234527890',
      address: indirizzo,
      hideContact: false,
    };
    const res: Response = await mutate(CREATE_USER, {
      variables: { auth: auth },
    });
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const token: string = res.data.createUser.response.token;
    await api.get(`/verify?id=${token}`).expect(200);
    expect(res.data.createUser.response.token).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data.createUser.response.user.isVerified).toBeFalsy();
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const meQuery = await query(ME);
    expect(meQuery).toMatchSnapshot();
    expect(fakeStart).toBeCalledTimes(1);
    expect(fakeStop).toBeCalledTimes(1);
  });

  it('query me is successfull if logged in, shows parameter meant for user Giovanni', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(ME_FULL);
    expect(res).toMatchSnapshot();
  });

  it('query single user is successfull if logged in, shows sensitive parameter meant for user in ctx', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const user = await User.findOne({ email: 'gio@prova.it' }).lean().exec();
    const res = await query(USER, {
      variables: {
        id: user?._id.toString(),
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('query single user is successfull if logged in, shows sensitive parameter meant for user in ctx and forbidden if not same user', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const user = await User.findOne({ email: 'gio2@prova.it' }).lean().exec();
    const res = await query(USER, {
      variables: {
        id: user?._id.toString(),
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('query users is successfull if logged in, shows sensitive parameters only to user in ctx', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(USERS);
    expect(res).toMatchSnapshot();
  });

  it('query me is successfull if logged in, shows parameter meant for user Mariuccio', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(ME_FULL);
    expect(res).toMatchSnapshot();
  });

  it('query me fails if not logged ', async () => {
    setOptions({
      request: {
        headers: {
          authorization: '123',
        },
      },
    });
    const res = await query(ME_FULL);
    expect(res).toMatchSnapshot();
  });

  it('query users fails if not logged ', async () => {
    setOptions({
      request: {
        headers: {
          authorization: '123',
        },
      },
    });
    const res = await query(USERS);
    expect(res).toMatchSnapshot();
  });

  it('create user mutation returns errors if parameters are invalid', async () => {
    const auth: UserInput = {
      email: 'sbagliata#oo.it',
      password: 'facile',
      firstName: 'Giova',
      lastName: 'Gibelli',
      pIva: '0101010101',
      phoneNumber: '123',
      address: indirizzo,
      hideContact: true,
    };
    const res = await mutate(CREATE_USER, {
      variables: { auth: auth },
    });
    expect(res).toMatchSnapshot();
    //expect(res.data.createUser.response).toBeNull();
    //expect(res.data.createUser.errors).toHaveLength(3);
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
