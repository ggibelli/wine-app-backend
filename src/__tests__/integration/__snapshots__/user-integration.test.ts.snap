// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test users create user mutation returns errors if parameters are invalid 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "errors": Array [
        Object {
          "name": "UserInputError",
          "text": "The email provided is not valid",
        },
        Object {
          "name": "UserInputError",
          "text": "The password provided is not valid",
        },
        Object {
          "name": "UserInputError",
          "text": "The PIVA provided is not valid",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test users create user mutation successfull and activation successfull 1`] = `
Object {
  "data": Object {
    "me": Object {
      "isVerified": true,
      "pIva": "01640250054",
      "phoneNumber": "1234527890",
    },
  },
}
`;

exports[`Integration test users invalid login mutation returns errors 1`] = `
Object {
  "data": Object {
    "login": Object {
      "errors": Array [
        Object {
          "name": "UserInputError",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test users query me fails if not logged  1`] = `
Object {
  "data": Object {
    "me": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "me",
      ],
    },
  ],
}
`;

exports[`Integration test users query me is successfull if logged in, shows parameter meant for user Giovanni 1`] = `
Object {
  "data": Object {
    "me": Object {
      "ads": Array [
        Object {
          "postedBy": Object {
            "firstName": "Giovanni",
          },
          "wineName": "Vino",
        },
      ],
      "isVerified": true,
      "messages": Array [
        Object {
          "content": "ciao",
          "sentBy": Object {
            "firstName": "Giovanni",
          },
          "sentTo": Object {
            "firstName": "Mariuccio",
          },
        },
        Object {
          "content": "ciao di nuovo",
          "sentBy": Object {
            "firstName": "Luigetto",
          },
          "sentTo": Object {
            "firstName": "Giovanni",
          },
        },
      ],
      "negotiations": Array [
        Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUserAd": Object {
            "firstName": "Mariuccio",
          },
        },
        Object {
          "createdBy": Object {
            "firstName": "Luigetto",
          },
          "forUserAd": Object {
            "firstName": "Giovanni",
          },
        },
      ],
      "pIva": "12345678901",
      "phoneNumber": "1234567890",
      "reviews": Array [
        Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUser": Object {
            "firstName": "Mariuccio",
          },
          "rating": "PERFECT",
        },
        Object {
          "createdBy": Object {
            "firstName": "Luigetto",
          },
          "forUser": Object {
            "firstName": "Giovanni",
          },
          "rating": "POOR",
        },
      ],
    },
  },
}
`;

exports[`Integration test users query me is successfull if logged in, shows parameter meant for user Mariuccio 1`] = `
Object {
  "data": Object {
    "me": Object {
      "ads": Array [
        Object {
          "postedBy": Object {
            "firstName": "Mariuccio",
          },
          "wineName": "Vino vendo",
        },
      ],
      "isVerified": true,
      "messages": Array [
        Object {
          "content": "ciao",
          "sentBy": Object {
            "firstName": "Giovanni",
          },
          "sentTo": Object {
            "firstName": "Mariuccio",
          },
        },
      ],
      "negotiations": Array [
        Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUserAd": Object {
            "firstName": "Mariuccio",
          },
        },
      ],
      "pIva": "12345678301",
      "phoneNumber": "1234567830",
      "reviews": Array [
        Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUser": Object {
            "firstName": "Mariuccio",
          },
          "rating": "PERFECT",
        },
      ],
    },
  },
}
`;

exports[`Integration test users query single user is successfull if logged in, shows sensitive parameter meant for user in ctx 1`] = `
Object {
  "data": Object {
    "user": Object {
      "ads": Array [
        Object {
          "postedBy": Object {
            "firstName": "Giovanni",
          },
          "wineName": "Vino",
        },
      ],
      "firstName": "Giovanni",
      "messages": Array [
        Object {
          "content": "ciao",
          "sentBy": Object {
            "firstName": "Giovanni",
          },
          "sentTo": Object {
            "firstName": "Mariuccio",
          },
        },
        Object {
          "content": "ciao di nuovo",
          "sentBy": Object {
            "firstName": "Luigetto",
          },
          "sentTo": Object {
            "firstName": "Giovanni",
          },
        },
      ],
      "negotiations": Array [
        Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUserAd": Object {
            "firstName": "Mariuccio",
          },
        },
        Object {
          "createdBy": Object {
            "firstName": "Luigetto",
          },
          "forUserAd": Object {
            "firstName": "Giovanni",
          },
        },
      ],
      "reviews": Array [
        Object {
          "createdBy": Object {
            "firstName": "Giovanni",
          },
          "forUser": Object {
            "firstName": "Mariuccio",
          },
          "rating": "PERFECT",
        },
        Object {
          "createdBy": Object {
            "firstName": "Luigetto",
          },
          "forUser": Object {
            "firstName": "Giovanni",
          },
          "rating": "POOR",
        },
      ],
    },
  },
}
`;

exports[`Integration test users query single user is successfull if logged in, shows sensitive parameter meant for user in ctx and forbidden if not same user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "ads": Array [
        Object {
          "postedBy": Object {
            "firstName": "Mariuccio",
          },
          "wineName": "Vino vendo",
        },
      ],
      "firstName": "Mariuccio",
      "messages": null,
      "negotiations": null,
      "reviews": null,
    },
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 12,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "user",
        "negotiations",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 20,
        },
      ],
      "message": "You can only see your own messages",
      "path": Array [
        "user",
        "messages",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 29,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "user",
        "reviews",
      ],
    },
  ],
}
`;

exports[`Integration test users query users fails if not logged  1`] = `
Object {
  "data": Object {
    "users": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "users",
      ],
    },
  ],
}
`;

exports[`Integration test users query users is successfull if logged in, shows sensitive parameters only to user in ctx 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "ads": Array [
          Object {
            "postedBy": Object {
              "firstName": "Mariuccio",
            },
            "wineName": "Vino vendo",
          },
        ],
        "email": "The user prefers to hide the email",
        "firstName": "Mariuccio",
        "messages": null,
        "negotiations": null,
        "phoneNumber": "The user prefers to hide the phone number",
        "reviews": null,
      },
      Object {
        "ads": Array [],
        "email": "The user prefers to hide the email",
        "firstName": "Luigetto",
        "messages": null,
        "negotiations": null,
        "phoneNumber": "The user prefers to hide the phone number",
        "reviews": null,
      },
      Object {
        "ads": Array [
          Object {
            "postedBy": Object {
              "firstName": "Giovanni",
            },
            "wineName": "Vino",
          },
        ],
        "email": "gio@prova.it",
        "firstName": "Giovanni",
        "messages": Array [
          Object {
            "content": "ciao",
            "sentBy": Object {
              "firstName": "Giovanni",
            },
            "sentTo": Object {
              "firstName": "Mariuccio",
            },
          },
          Object {
            "content": "ciao di nuovo",
            "sentBy": Object {
              "firstName": "Luigetto",
            },
            "sentTo": Object {
              "firstName": "Giovanni",
            },
          },
        ],
        "negotiations": Array [
          Object {
            "createdBy": Object {
              "firstName": "Giovanni",
            },
            "forUserAd": Object {
              "firstName": "Mariuccio",
            },
          },
          Object {
            "createdBy": Object {
              "firstName": "Luigetto",
            },
            "forUserAd": Object {
              "firstName": "Giovanni",
            },
          },
        ],
        "phoneNumber": "1234567890",
        "reviews": Array [
          Object {
            "createdBy": Object {
              "firstName": "Giovanni",
            },
            "forUser": Object {
              "firstName": "Mariuccio",
            },
            "rating": "PERFECT",
          },
          Object {
            "createdBy": Object {
              "firstName": "Luigetto",
            },
            "forUser": Object {
              "firstName": "Giovanni",
            },
            "rating": "POOR",
          },
        ],
      },
      Object {
        "ads": Array [],
        "email": "giollo@ollo.it",
        "firstName": "Giova",
        "messages": null,
        "negotiations": null,
        "phoneNumber": "1234527890",
        "reviews": null,
      },
    ],
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 14,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "users",
        0,
        "negotiations",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 22,
        },
      ],
      "message": "You can only see your own messages",
      "path": Array [
        "users",
        0,
        "messages",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 31,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "users",
        0,
        "reviews",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 14,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "users",
        1,
        "negotiations",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 22,
        },
      ],
      "message": "You can only see your own messages",
      "path": Array [
        "users",
        1,
        "messages",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 31,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "users",
        1,
        "reviews",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 14,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "users",
        3,
        "negotiations",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 22,
        },
      ],
      "message": "You can only see your own messages",
      "path": Array [
        "users",
        3,
        "messages",
      ],
    },
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 31,
        },
      ],
      "message": "You can only see your own negotiations",
      "path": Array [
        "users",
        3,
        "reviews",
      ],
    },
  ],
}
`;
